### DISCUSSION TITLE
Shadowed unused import not detected when using `from __future__ import annotations`

### DISCUSSION THREAD

**Discussion by injust** [microsoft/pyright#9640] [Q&A] [answered]
```python
from __future__ import annotations

from enum import Enum

from packaging.version import Version


class Foo:
    class Version(Enum):
        pass

    version: Version | None = None
```

Inside `Foo`, the unused `from packaging.version import Version` import is shadowed by the `Version` enum. But pyright doesn't detect the unused import, unless I remove `from __future__ import annotations`.

Is this something that pyright could catch? I don't really understand why `from __future__ import annotations` is affecting the detection of unused imports.

Edit: I think this is actually a bug. Pyflakes detects the unused import with/without `from __future__ import annotations`.

**Comment by erictraut** [comment, marked as answer]
Pyright is working correctly here. The `from __future__ import annotations` directive tells the runtime to defer evaluation of type annotations. When the annotation `Version | None` is evaluated in a deferred manner, the imported `Version` is used. If you comment out the `from __future__ import annotations`, the annotation is evaluated at runtime immediately when the runtime encounters that statement, and at that point `Version` refers to the inner class of that name, and the imported symbol of that name is not referenced.

**Reply by injust** [reply]
I see, thanks for the explanation.
